mono/metadata/pal-icalls.c

```diff
@@ -0,0 +1,102 @@
+/**
+ * \file
+ * System.Native PAL internal calls
+ *
+ * Copyright 2018 Microsoft
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+*/
+
+#include <config.h>
+#include <glib.h>
+#include <mono/metadata/handle.h>
+#include <mono/utils/hazard-pointer.h>
+#include <mono/utils/mono-threads.h>
+
+#include "pal-icalls.h"
+
+#if !defined(HOST_WIN32) && !defined(HOST_WASM) /* Unix System.Native PAL */
+
+extern int32_t SystemNative_Read(intptr_t fd, void* buffer, int32_t bufferSize);
+
+static void
+interrupt_read (gpointer data)
+{
+	MonoNativeThreadId tid = (MonoNativeThreadId) data;
+	MonoThreadHazardPointers *hp;
+	MonoThreadInfo *info;
+
+	if (tid == mono_native_thread_id_get ())
+		return;
+
+	mono_thread_info_suspend_lock ();
+	hp = mono_hazard_pointer_get ();
+	info = mono_thread_info_lookup (tid);
+	if (!info) {
+		mono_thread_info_suspend_unlock ();
+		return;
+	}
+	mono_threads_begin_global_suspend ();
+	mono_threads_suspend_abort_syscall (info);
+	mono_threads_wait_pending_operations ();
+
+	mono_hazard_pointer_clear (hp, 1);
+	mono_threads_end_global_suspend ();
+	mono_thread_info_suspend_unlock ();
+}
+
+gint32
+ves_icall_CoreFX_Interop_Sys_Read (MonoSafeHandleHandle fd_handle, gchar* buffer, gint32 count)
+{
+	gint32 result;
+	gpointer fd = MONO_HANDLE_GETVAL(fd_handle, handle);
+	gboolean interrupted;
+
+	do {
+		mono_thread_info_install_interrupt (interrupt_read, (gpointer) (gsize) mono_native_thread_id_get (), &interrupted);
+		if (interrupted)
+			return -1;
+
+		MONO_ENTER_GC_SAFE;
+		result = SystemNative_Read ((intptr_t) fd, buffer, count);
+		MONO_EXIT_GC_SAFE;
+		mono_thread_info_uninstall_interrupt (&interrupted);
+	} while (!interrupted && result < 0 && errno == EINTR);
+
+	if (interrupted)
+		return -1;
+
+	return result;
+}
+
+#if defined(__APPLE__)
```

Comment:
@luhenry @kumpera do we want to mix unix and apple code in the same file?

Reply:
That's fine as long as the code is not overblown with preprocessor conditionals. In this case, it's definitely manageable.